include/mqt-core/UndirectedGraph.hpp

```diff
@@ -0,0 +1,181 @@
+//
+// This file is part of the MQT QMAP library released under the MIT license.
+// See README.md or go to https://github.com/cda-tum/mqt-qmap for more
+// information.
+//
+
+#pragma once
+
+#include "Definitions.hpp"
+
+#include <numeric>
+#include <sstream>
+#include <unordered_set>
+
+namespace qc {
+
+template <class V, class E> class UndirectedGraph {
+  using F = std::shared_ptr<E>;
+
+protected:
+  // the adjecency matrix works with indices
+  std::vector<std::vector<F>> adjacencyMatrix{};
```

Comment:
Yes, the directed graphs could be augmented with an edge type as well. I simply did not do it because I do not need it. I would follow a demand-driven approach here and add it whenever it is needed. What do you think?

Reply:
I like the suggestion with optional and adopted it.